File: Histograms.m
%% Energy Histogram

%function [EnergySpectrumNtuple, EnergySpectrumHisto, bin_centers, num_bins, x_min, x_max] = Histograms
function [EnergySpectrumHisto, bin_centers, num_bins, x_min, x_max] = Histograms

% Define file names
%fileNameNtuple = 'MC-ADD_Results_nt_Photons.csv';
fileNameHisto  = 'MC-ADD_Results_h1_Energy_Deposit.csv';

% Initialize empty arrays
%EnergyNtuple = [];
EnergyHisto  = [];

try

    % % ::: Read Ntuple File :::
    % if exist(fileNameNtuple, 'file') == 2
    % 
    %     DataNtuple = readtable(fileNameNtuple);
    %     EnergyNtuple = table2array(DataNtuple(:,5));
    % 
    % else
    %     warning("Ntuple file not found: %s", fileNameNtuple);
    % end


    % ::: Read Histogram File :::
    if exist(fileNameHisto, 'file') == 2

        % The first 6 lines are Geant4 metadata and line 7
        % contains the CSV column names.
        % Numerical data therefore begin on line 8.
        DataHisto = readmatrix(fileNameHisto, 'NumHeaderLines', 7);

        % Column 1 contains histogram entries
        EnergyHisto = DataHisto(:,1);

        % Remove Geant4 underflow and overflow bins
        EnergyHisto = EnergyHisto(2:end-1);

    else
        warning("Histogram file not found: %s", fileNameHisto);
    end


catch ME
    rethrow(ME);
end


% ::: Clean spectra :::
%EnergySpectrumNtuple = EnergyNtuple(EnergyNtuple > 0);
EnergySpectrumHisto  = EnergyHisto;


% :::::: Define histogram bins ::::::

RunActionFile = 'src/RunAction.cc';

fid = fopen(RunActionFile, 'r');

if fid == -1
    error('Error opening the file. Check the path.');
end

lines = textscan(fid, '%s', 'Delimiter', '\n');
fclose(fid);

lines = lines{1};

HistogramInfo = lines{20};

tokens = regexp(HistogramInfo, '(\d+),\s*([\d.]+),\s*(\d+)', 'tokens');

if isempty(tokens)
    error('Could not extract histogram information. Check the format.');
end

HistInfo = str2double(tokens{1});

num_bins = HistInfo(1);
x_min    = HistInfo(2);
x_max    = HistInfo(3);


% ::: Define bin centers :::

bin_edges = linspace(x_min, x_max, num_bins + 1);

bin_centers = (bin_edges(1:end-1) + bin_edges(2:end)) / 2;


% Force both arrays to columns
EnergySpectrumHisto = EnergySpectrumHisto(:);
bin_centers = bin_centers(:);


% ::: Debugging output :::

% fprintf('\nHistogram diagnostic:\n');
% fprintf('  Raw Geant4 rows after readmatrix = %d\n', size(DataHisto,1));
% fprintf('  Physical histogram bins         = %d\n', length(EnergySpectrumHisto));
% fprintf('  Expected bins from RunAction.cc = %d\n', num_bins);
% fprintf('  Bin centers                     = %d\n\n', length(bin_centers));


% ::: Consistency check :::

if length(EnergySpectrumHisto) ~= length(bin_centers)

    error("Mismatch: bin_centers has %d elements but EnergySpectrumHisto has %d elements.", length(bin_centers), length(EnergySpectrumHisto));

end

end

File: Histograms.py
import numpy as np
import pandas as pd
import re


def Histograms():

    # Define file names
    fileNameHisto = 'MC-ADD_Results_h1_Energy_Deposit_FINAL.csv'

    # Initialize empty array
    EnergyHisto = np.array([])

    try:

        # The first 6 lines are Geant4 metadata and line 7
        # contains the CSV column names.
        # Numerical data therefore begin on line 8.
        DataHisto = pd.read_csv(fileNameHisto, skiprows=7, header=None)

        # Column 1 contains histogram entries
        EnergyHisto = DataHisto.iloc[:, 0].to_numpy(dtype=float)

        # Remove Geant4 underflow and overflow bins
        EnergyHisto = EnergyHisto[1:-1]

    except FileNotFoundError:
        raise FileNotFoundError(
            f'Histogram file not found: {fileNameHisto}'
        )

    except Exception as e:
        raise RuntimeError(
            f'Error reading histogram file: {e}'
        )


    EnergySpectrumHisto = EnergyHisto


    # :::::: Define histogram bins ::::::

    RunActionFile = 'src/RunAction.cc'

    try:
        with open(RunActionFile, 'r') as file:
            lines = file.readlines()

    except FileNotFoundError:
        raise FileNotFoundError(
            f'Error opening the file: {RunActionFile}'
        )


    # Histogram information is located in line 20 of RunAction.cc
    # Python indexing starts at 0
    HistogramInfo = lines[19]

    tokens = re.search(
        r'(\d+),\s*([\d.]+),\s*([\d.]+)',
        HistogramInfo
    )

    if tokens is None:
        raise ValueError(
            'Could not extract histogram information. Check the format.'
        )


    num_bins = int(tokens.group(1))
    x_min    = float(tokens.group(2))
    x_max    = float(tokens.group(3))


    # ::: Bin centers :::

    bin_edges = np.linspace(
        x_min,
        x_max,
        num_bins + 1
    )

    bin_centers = (
        bin_edges[:-1] + bin_edges[1:]
    ) / 2


    # ::: Debugging: Check if lengths match :::

    if len(EnergySpectrumHisto) != len(bin_centers):
        raise ValueError(
            f'Mismatch: bin_centers has {len(bin_centers)} elements '
            f'but EnergySpectrumHisto has {len(EnergySpectrumHisto)} elements.'
        )


    return (
        EnergySpectrumHisto,
        bin_centers,
        num_bins,
        x_min,
        x_max
    )